WDL/runtime/workflow.py

```diff
@@ -589,41 +594,59 @@ def run_local_workflow(
-                    run_callee = run_local_workflow
-                else:
-                    assert False
-                _, outputs = run_callee(
-                    next_call.callee,  # pyre-fixme
-                    next_call.inputs,
-                    run_id=next_call.id,
-                    run_dir=os.path.join(run_dir, next_call.id),
-                )
-                state.call_finished(next_call.id, outputs)
-    except Exception as exn:
-        logger.debug(traceback.format_exc())
-        if isinstance(exn, TaskFailure):
-            logger.error("%s failed", getattr(exn, "run_id"))
-        else:
-            msg = ""
-            if hasattr(exn, "job_id"):
-                msg += getattr(exn, "job_id") + " "
-            msg += exn.__class__.__name__
-            if str(exn):
-                msg += ", " + str(exn)
-            logger.error(msg)
-            logger.info("run directory: %s", run_dir)
-        raise
+    with TerminationSignalFlag(logger) as terminating:
+        try:
+            with thread_pool as executor:
+                while state.outputs is None:
+                    if _test_pickle:
+                        state = pickle.loads(pickle.dumps(state))
+                    next_call = state.step()
+                    while next_call:
+                        if isinstance(next_call.callee, Tree.Task):
+                            run_callee = run_local_task
+                        elif isinstance(next_call.callee, Tree.Workflow):
+                            run_callee = run_local_workflow
+                        else:
+                            assert False
+                        future = executor.submit(run_callee,
+                                                 next_call.callee,  # pyre-fixme
+                                                 next_call.inputs,
+                                                 run_id=next_call.id,
+                                                 run_dir=os.path.join(run_dir, next_call.id),
+                                                 thread_pool=thread_pool,
+                                                 )
+                        future_task_map[future] = next_call.id
+                        next_call = state.step()
+                    done_iter = futures.as_completed(future_task_map)
+                    future = next(done_iter, None)
+                    if future:
+                        try:
+                            _, outputs = future.result()
+                            call_id = future_task_map[future]
+                            state.call_finished(call_id, outputs)
+                            future_task_map.pop(future)
+
+                        except Exception as e:
+                            for key in future_task_map:
+                                key.cancel()
+                            pid = os.getpid()
+                            os.kill(pid, signal.SIGTERM)
```

Comment:
We can outdent this wind-down logic to put it alongside the other exception handler below. It's possible for `state.step()` to raise an exception if e.g. something goes wrong evaluating one of the user's WDL expressions, and we should try to clean up in any such case.

Reply:
Also probably worth some comments to explain what we mean to happen!

Reply:
👍